### PR TITLE
LocalNames and NameMap's Marshal Function Bug Fix

### PR DESCRIPTION
**Pull request by zhangzhiqiangcs**
I use NameSection in my project. 
I think NameMap and LocalNames' Marshal function miss out size information constrasted to it's Unmarshal function.

### CODE REVIEW

**Comment by sbinet** [comment]
# [Codecov](https://codecov.io/gh/go-interpreter/wagon/pull/187?src=pr&el=h1) Report
> Merging [#187](https://codecov.io/gh/go-interpreter/wagon/pull/187?src=pr&el=desc) into [master](https://codecov.io/gh/go-interpreter/wagon/commit/8dd99d5adacc2fcb2a87b96c6001ecc84a6ed09c&el=desc) will **decrease** coverage by `0.05%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/go-interpreter/wagon/pull/187?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #187      +/-   ##
==========================================
- Coverage   69.77%   69.72%   -0.06%     
==========================================
  Files          49       49              
  Lines        5495     5499       +4     
==========================================
  Hits         3834     3834              
- Misses       1318     1322       +4     
  Partials      343      343              
```



------

[Continue to review full report at Codecov](https://codecov.io/gh/go-interpreter/wagon/pull/187?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/go-interpreter/wagon/pull/187?src=pr&el=footer). Last update [8dd99d5...614874e](https://codecov.io/gh/go-interpreter/wagon/pull/187?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by sbinet** [comment]
thanks. could you add a few tests so we don't introduce regressions unadvertantly?

thanks again.

**Comment by zhangzhiqiangcs** [comment]
> thanks. could you add a few tests so we don't introduce regressions unadvertantly?
> 
> thanks again.

No problem, I'll do it later.

**Comment by twitchyliquid64** [comment]
Thanks for writing this! I've double-checked the spec, this seems correct. No idea how this worked in the first place!

File: wasm/section_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 The go-interpreter Authors.  All rights reserved.
```
Comment (sbinet) [inline wasm/section_test.go]:
```suggestion
// Copyright 2020 The go-interpreter Authors.  All rights reserved.
```

File: wasm/section_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			t.Fatal("can not find name custom section")
+		}
+
+		var nSec wasm.NameSection
+		err = nSec.UnmarshalWASM(bytes.NewReader(nameCustom.Data))
+		if err != nil {
+			t.Fatalf("error name Section Unmarshal %v", err)
+		}
+
+		if len(nSec.Types[wasm.NameFunction]) > 0 {
```
Comment (sbinet) [inline wasm/section_test.go]: please invert the boolean test:
```suggestion
		if len(nSec.Types[wasm.NameFunction]) == 0 {
		    t.Fatalf(...)
		}
```

File: wasm/section_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+				t.Fatalf("error FunctionNames Marshal %v", err)
+			}
+			if !bytes.Equal(buf.Bytes(), nSec.Types[wasm.NameFunction]) {
+				t.Fatal("error Marshal and Unmarshal FunctionNames")
+			}
+		} else {
+			t.Fatalf("%s doesn't have custom FunctionNames section", fname)
+		}
+
+		if len(nSec.Types[wasm.NameLocal]) > 0 {
```
Comment (sbinet) [inline wasm/section_test.go]: ditto.

**Comment by zhangzhiqiangcs** [comment]
> Thanks for writing this! I've double-checked the spec, this seems correct. No idea how this worked in the first place!

In fact, the module's encode and decode function don't Marshal and Unmarshal NameSubSection's data. It matters when you modified the NameSubSection and want to write back to the Module.

**Comment by zhangzhiqiangcs** [comment]
> just a few nitpicks. to address.
> 
> could you also send a PR against go-interpreter/license, adding yourself to the `AUTHORS` and `CONTRIBUTORS` files?

I have done this in the pr https://github.com/go-interpreter/license/pull/19. Thanks a lot~

**Pull request merged**